src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -780,43 +780,28 @@ private void TrackNullableStateForAssignment(BoundExpression value, TypeSymbolWi
 
                 InheritDefaultState(targetSlot);
 
-                if (targetType.IsReferenceType)
-                {
-                    // https://github.com/dotnet/roslyn/issues/31395,
-                    // https://github.com/dotnet/roslyn/issues/29968 We should copy all tracked state from `value`,
-                    // regardless of BoundNode type, but we'll need to handle cycles. (For instance, the
-                    // assignment to C.F below. See also NullableReferenceTypesTests.Members_FieldCycle_01.)
-                    // class C
-                    // {
-                    //     C? F;
-                    //     C() { F = this; }
-                    // }
-                    // For now, we copy a limited set of BoundNode types that shouldn't contain cycles.
-                    if ((value.Kind == BoundKind.ObjectCreationExpression || value.Kind == BoundKind.AnonymousObjectCreationExpression || value.Kind == BoundKind.DynamicObjectCreationExpression || targetType.TypeSymbol.IsAnonymousType) &&
-                        areEquivalentTypes(targetType, valueType)) // https://github.com/dotnet/roslyn/issues/29968 Allow assignment to base type.
+                if (slotWatermark < 0)
+                {
+                    slotWatermark = GetSlotWatermark();
+                }
+
+                if (areEquivalentTypes(targetType, valueType)) // https://github.com/dotnet/roslyn/issues/29968 Allow assignment to base type.
```

Comment:
I have trouble understanding the purpose of this `areEquivalentTypes` call. 
If I remove it, no test seems to be affected.
Could you clarify purpose and/or add test coverage to should usefulness?

Reply:
The `areEquivalentTypes` call will fail when `valueType` is null or when `valueType` contains error types. I don't have a scenario yet where the check makes a difference though.

---
In reply to: [257339664](https://github.com/dotnet/roslyn/pull/33395#discussion_r257339664) [](ancestors = 257339664)

Reply:
`areEquivalentTypes` only fails when `valueType` is `default` or when `targetType` or `valueType` contain error types. Removing the check does not change the behavior but I'll leave this check here until we're ensuring we're not copying state for invalid members. See [https://github.com/dotnet/roslyn/issues/33428](https://github.com/dotnet/roslyn/issues/33428).

---
In reply to: [257365869](https://github.com/dotnet/roslyn/pull/33395#discussion_r257365869) [](ancestors = 257365869,257339664)